### ISSUE TITLE
Plural Resource Names

### ISSUE THREAD

**Issue by mhluska** [mrichar1/jsonapi-vuex#19]
Does this library consider singular vs plural resource names? For example, when I try to update a resource:

```js
this.$store.dispatch('jv/patch', this.todo);
```

It will send a request to /api/v1/**todo**/24975302, however Rails has defined my endpoint as /api/v1/**todos**/24975302.

I noticed in your code that it might be possible to pass a string instead of a record, so I tried to specify a plural path like so:

```js
this.$store.dispatch('jv/patch', `todos/${this.todo._jv.id}/`);
```

But this throws an error:

```
vue.runtime.esm.js?2b0e:1887 TypeError: Cannot use 'in' operator to search for '_jv' in todos/24975302/
    at normToJsonapi (jsonapi-vuex.js?1c7f:380)
    at Store.patch (jsonapi-vuex.js?1c7f:163)
    at Array.wrappedActionHandler (vuex.esm.js?2f62:721)
    at Store.dispatch (vuex.esm.js?2f62:428)
    at Store.boundDispatch [as dispatch] (vuex.esm.js?2f62:322)
    at VueComponent.completeTodo (DashboardTodo.vue?d0b8:37)
    at invokeWithErrorHandling (vue.runtime.esm.js?2b0e:1854)
    at HTMLButtonElement.invoker (vue.runtime.esm.js?2b0e:2178)
    at HTMLButtonElement.original._wrapper (vue.runtime.esm.js?2b0e:6876)
```

Currently I am working around this issue using this hack:

```js
methods: {
  completeTodo() {
    this.todo._jv.type = 'todos';
    this.todo['is-done'] = true;     
    this.$store.dispatch('jv/patch', this.todo);
    this.todo._jv.type = 'todo';
  },
},
```

**Comment by mrichar1** [comment]
There's currently no support for pluralization in the library.

In general, it's a bad idea to use differently named collection/item endpoints for data of the same jsonapi type, since it makes discoverability more difficult, and requires extra work on the client-side to deal with.
So if you are in control of the API server I'd look into forcing it to use the same name for both.

However this isn't an option when using other people's APIs so this is probably a feature that will need adding.

The simplest approach is to have a `name` function that takes the `singular` form, and if an alternative `collection` name exists, return that.

There are specialist libraries that do 'clever' pluralisation such as [pluralize](https://www.npmjs.com/package/pluralize) - however I'm reluctant to pull in a fully-featured library as a dependency for what should hopefully not be a common use case (and the simple function described above can easily be replaced by anyone using this library to be more complex if needed).

**Comment by mrichar1** [comment]
Looking at this again, I now realise that there's actually 2 different potential naming issues. The first is singular vs plural (for item vs colletion), but the 2nd is where the resource type and the endpoint differ.

I'll need to implement some kind of matrix to cover all possible outcomes:

1. Type: blog, item: blog, collection: blog  # All match (current approach)
2. Type: blog, item: blog, collection: blogs # Plural collection endpoint
3. Type: article, item: blog, collection: blog # EP and type differ, no plurals.
4. Type: article, item: blog, collection: blogs # EP and type differ, with plurals.

Again, I'll start on this once the `status` branch work is merged.

**Comment by stefanvanherwijnen** [comment]
I think this is related to this issue:
```$store.dispatch('jv/get', 'users')``` will ```GET``` the '/users' endpoint, however, I can only retrieve the results with ```$store.getters['jv/get']('user')```.
I haven't looked into the code, but I suspect it is looking up the data with ```type```user. So defining 'users' instead will return no results.

I think you are right that this is more complex than it seems. As far as I know, plural endoints are correct. However, if an object has a relationship of one user, the relationship name should be singular. And if the user is the owner of the object, the relationship should be called ```owner```.

Other json-api packages (e.g. https://github.com/danivek/json-api-serializer) require you to define the relationships yourself, which does tackle this problem.

**Comment by mrichar1** [comment]
@stefanvanherwij yes this is the 'reverse' of how I've been mostly describing the situation so far - and yes, it really relies on the person using the endpoint to know that `/users` returns objects of type `user`.

#54 allows overriding the endpoint name for a given type when calling actions, which I _think_ is enough to make APIs with differing type/EP names useable - i.e you just need to define a mapping in your code, then make sure that the `axios `url` param is set by a function that reads the mapping.

I'd be reluctant to jump to implementing registration of type<->ep relationship mappings in this module, or at least if this was added make it an optional extra on top of the default approach and existing workarounds. I know one of the reasons I steered away from using existing serialisation libraries was the need to reimplement the whole api schema in the client code before even starting to access it!

Always happy to hear comments and suggestions on how to improve things, and accept Issues or PRs which steer things in new directions if they are useful - it's often difficult to know all the corner cases that others will hit when using your code! :smiley:

**Comment by mrichar1** [comment]
Fixed by #54
